### PR TITLE
Creates DiscreteMaterial object to enable performing iterative proper…

### LINKED ISSUE #5166: Add ability to recompute Materials at a quadrature point
`Material` will add the following methods:
- `computeProperty(unsigned int qp)` - this will set _qp to qp and call `computeQpProperties()
- `unsigned int id materialProperty(std::string prop_name)` - this will take a material property name and return a unique ID for the property name
- `recomputeMaterial(unsigned int id, unsigned int qp)` recomputes the necessary `Material` objects
at the provided quadrature point

### PR DESCRIPTION
**Pull request by aeslaughter**
This is nearly the same as my previous attempt to have materials call compute on other materials, but the calling of the recompute method is explicitly performed by retrieving a special type of Material object, a DiscreteMaterial, which is in keeping with the recently added DiscreteElementUserObject. This design was suggested by @permcody and @sapitts.

(closes #5166)

### CODE REVIEW

**Comment by aeslaughter** [comment]
This is ready to go, I have a YAK patch waiting. 

I would like to get this in soon, @sapitts has been waiting patiently for me to get this finished up.

Just as a note, I started working on a branch of this that only had two storage containers (`Material` and `DiscreteMaterial`) rather than also having a `MaterialBase` storage container. However, doing this added a lot of code because you have to integrate the two lists an about a half a dozen places, to me just storing a third container is worth it to avoid this merging of two structures.

**Comment by sapitts** [comment]
pinging @jasondhales and @bwspenc in case they have comments on the discrete (constitutive) material.  Thank you @aeslaughter for creating this new material type.

**Comment by bwspenc** [comment]
In my very quick look at this, this seems like exactly what we need.  Thanks!

One question -- what does the name "Discrete" mean in the context of the discrete materials and user objects?

**Comment by aeslaughter** [comment]
I choose "Discrete" to be consistent with Discrete UserObject, which is
similar in concept. I am not sure why they selected "discrete" originally,
I think it was the best name we could think of.

**Comment by bwspenc** [comment]
@aeslaughter So does Discrete just mean something that MOOSE doesn't call automatically?  I'm not saying it should be changed or anything, just trying to understand what is special about an object with "Discrete" in its name.

**Comment by aeslaughter** [comment]
On Thu, Mar 3, 2016 at 3:03 PM, Ben Spencer <notifications@github.com>
wrote:

> @aeslaughter
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_aeslaughter&d=BQMCaQ&c=54IZrppPQZKX9mLzcGdPfFD1hxrcB__aEkJFOKJFd00&r=h7heP8xwI1i_HikChvhFbEBurKirgfOCdwgBxB9lM8c&m=qFomJKxDnuzXC37XmcqAdnb84g_RNRSnPIbeo-bU0I8&s=K2L7WTDa2H5y4SlRPskYQt3iJn6Ryv2PMcmjG-mtoUo&e=>
> So does Discrete just mean something that MOOSE doesn't call automatically?
>
Yes

File: test/include/materials/RecomputeMaterial.h [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  RecomputeMaterial(const InputParameters & parameters);
+
+protected:
+  void computeQpProperties();
+  void resetQpProperties();
+
+private:
+  const VariableValue & _var;
+  MaterialProperty<Real> & _f;
+  MaterialProperty<Real> &_f_prime;
```
Comment (jwpeterson) [inline test/include/materials/RecomputeMaterial.h]: Need a space after `&`

File: test/src/materials/RecomputeMaterial.C [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  _f[_qp] = 84;
+  _f_prime[_qp] = 42;
+}
+
+void
+RecomputeMaterial::computeQpProperties()
+{
+  Real x = _p[_qp];
+  _f[_qp] = x * x - _var[_qp];
+  _f_prime[_qp] = _var[_qp]*x;
```
Comment (jwpeterson) [inline test/src/materials/RecomputeMaterial.C]: Double check the math on this?  Seems like the derivative should be `2*x`...

File: test/src/materials/NewtonMaterial.C [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+{
+  _p[_qp] = 0.5; // initial guess
+
+  // Newton iteration for find p
+  for (unsigned int i = 0; i < _max_iterations; ++i)
+  {
+    _discrete.computeProperties(_qp);
+    if (std::abs(_f[_qp]) < _tol)
+      break;
+    _p[_qp] = _p[_qp] - _f[_qp] / _f_prime[_qp];
```
Comment (jwpeterson) [inline test/src/materials/NewtonMaterial.C]: This would be a good place to use `-=`

File: test/include/materials/NewtonMaterial.h [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  NewtonMaterial(const InputParameters & parameters);
+  virtual ~NewtonMaterial(){};
+
+protected:
+  void computeQpProperties();
+
+private:
+  const Real & _tol;
+  const MaterialProperty<Real> & _f;
+  const MaterialProperty<Real> &_f_prime;
```
Comment (jwpeterson) [inline test/include/materials/NewtonMaterial.h]: Space after `&`

File: test/src/materials/NewtonMaterial.C [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+#include "NewtonMaterial.h"
+#include "DiscreteMaterial.h"
+
+
+template<>
+InputParameters validParams<NewtonMaterial>()
+{
+  InputParameters params = validParams<Material>();
+  params.addRequiredParam<std::string>("f_name", "The name of the property that holds the value to of the function for which the root is being computed");
+  params.addRequiredParam<std::string>("f_prime_name", "The name of the property that holds the value to of the derivative of the function");
```
Comment (dschwen) [inline test/src/materials/NewtonMaterial.C]: _ holds the value to of_ ?

This class could be useful in the framework. You could use the derivative material stuff for that.

Comment (aeslaughter) [inline test/src/materials/NewtonMaterial.C]: I will fix the typo, it is just to demonstrate the functionality. I think it is too limited to be in the framework, but if you want to generalize, go nuts after this is in.

File: test/src/materials/RecomputeMaterial.C [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+  _f[_qp] = 84;
+  _f_prime[_qp] = 42;
+}
+
+void
+RecomputeMaterial::computeQpProperties()
+{
+  Real x = _p[_qp];
+  _f[_qp] = x * x *_var[_qp];
+  _f_prime[_qp] = 2*_var[_qp]*x;
```
Comment (jwpeterson) [inline test/src/materials/RecomputeMaterial.C]: Space around operators on lines 50 and 51.

**Comment by jwpeterson** [comment]
I think I'm done looking at this, but it's a huge single commit so I'm sure I missed something.

**Comment by jasondhales** [comment]
I agree with Ben, this looks like what we need.  Thanks Andrew and Stephanie for pushing this through.

I look forward to trying it out.

**Comment by aeslaughter** [comment]
I need to do a YAK patch to get this through, so merge should wait until
Monday.

**Comment by friedmud** [comment]
Just one question: I thought we created DiscreteUserObjects so that we
wouldn't need to do this sort of thing.  Do we really need both mechanisms?

Derek

**Comment by bwspenc** [comment]
@friedmud The main reason for this is that these material models need to register, read, and write material properties that are specific to these models. With DiscreteUserObjects, the code is pretty ugly because you have to have the containing material model register the material properties for the submodels, and you also have to pass a set of properties into them. This should make the architecture much more cleaner.

**Comment by friedmud** [comment]
Ok - I was originally for allowing re-calculation of Material properties
(we worked on that some last summer in the tiger team).  But I thought we
were going to go the DiscreteUserObject direction instead.

So now my question is: should we remove DiscreteUserObject?

**Comment by dschwen** [comment]
I'm at lunch now, but let's not jump to conclusions just now. I'll write
what I use the user object for later this afternoon. It could technically
be accomplished by a material, but the flexible interface makes the design
changed IMO.

**Comment by permcody** [comment]
image

**Comment by aeslaughter** [comment]
@dschwen Do you want to add why we need to keep DiscreteUO?
@permcody Do you want to weight in on this PR, you were apart of the re-design?

I want to get this merged today if we can agree.

**Comment by dschwen** [comment]
Ok, I'm using the DiscreteUserObject to allow for a pluggable system that generates a list of recoil atoms at each element that is contributed to a global list. I'd like the user to be able to specify a set of those and I like being able to specify my own interface for those (i.e. a method that appends an arbitrary number of entries entries to an existing vector). I could make this work with vector material properties (and copying them around) but I'd end up having more complicated/awkward code.
DiscreteUO is not a large system at all (compared to the recalculatable material properties) so if this isn't a big deal I'd like to keep it around.

File: framework/include/base/DependencyResolverInterface.h [hunk 7]
```diff
@@ -23,6 +23,12 @@
 
 // MOOSE includes
 #include "DependencyResolver.h"
+#include "InputParameters.h"
+
+// Forward Declarations
+class DependencyResolverInterface;
+
+
```
Comment (permcody) [inline framework/include/base/DependencyResolverInterface.h]: Looks like you left a little bit of stuff from your previous attempt. You probably don't need to touch this file at all. The remaining three changes are all whitespace.

File: framework/include/materials/Material.h [hunk 8]
```diff
@@ -15,381 +15,35 @@
...
-#include "TransientInterface.h"
-#include "MaterialPropertyInterface.h"
-#include "PostprocessorInterface.h"
-#include "DependencyResolverInterface.h"
-#include "Restartable.h"
-#include "ZeroInterface.h"
-#include "MeshChangedInterface.h"
-#include "OutputInterface.h"
-#include "RandomInterface.h"
+// MOOSE includes
```
Comment (permcody) [inline framework/include/materials/Material.h]: Why is this line here? Are they all MOOSE includes?

File: framework/src/materials/MaterialBase.C [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+}
+
+
+void
+MaterialBase::registerPropName(std::string prop_name, bool is_get, MaterialBase::Prop_State state)
+{
+  if (!is_get)
+  {
+    _props_to_flags[prop_name] |= static_cast<int>(state);
+    if (static_cast<int>(state) % 2 == 0)
```
Comment (permcody) [inline framework/src/materials/MaterialBase.C]: I guess this is old code (just moved) but somebody went through the trouble of using a bitfield here and even "or'd" things together but then did a mathematical operator instead of a bit operator on this line. It might have been me though so I guess I won't poke too much fun.

File: framework/include/materials/MaterialBase.h [hunk 10]
```diff
@@ -0,0 +1,373 @@
...
+   * Initialize stateful properties at quadrature points.  Note when using this function you only need to address
+   * the "current" material properties not the old ones directly, i.e. if you have a property named "_diffusivity"
+   * and an older property named "_diffusivity_old".  You only need to initialize diffusivity.  MOOSE will use
+   * copy that initial value to the old and older values as necessary.
+   */
+  virtual void initQpStatefulProperties();
+
+  /**
+   * Compute material properties in quadrature point
+   * Materials has to redefine this.
```
Comment (permcody) [inline framework/include/materials/MaterialBase.h]: redefine -> define

**Comment by permcody** [comment]
:+1:

**Comment by dschwen** [comment]
After reading the announcement and the discussion on the mailing list I do not understand why this system was implemented as a Material and what its advantages over the Discrete User Objects are. Unlike regular materials which are fetched based on material _property_ names, and which can use the block restriction system, the discrete materials are fetched via _material_ names and cannot be set up in a per-block fashion. I think this is confusing and I don't see the point of using a "Materials" system for the whole on-the-fly recalculation anymore. Maybe somebody can make me see the light...

**Comment by aeslaughter** [comment]
It exists for one reason: to recompute MaterialProperties on demand, this
cannot be done otherwise.

DiscreteMaterial objects are exactly the same as Material object, except
the user calls compute rather than MOOSE. They can't be restricted to
block/boundary because it doesn't make any sense, the user is calling
compute, which can be done from anything inheriting from
MaterialPropertyInterface. So, if you want to "block restrict" the
DiscreteProperties then call the compute from a block restricted object.

My test does exactly this, on onside of the domain two material properties
are calculated by GenericConstantMaterial on the other side they are
defined by a DiscreteMaterial. This DiscreteMaterial is computed by another
material (NewtonMaterial) that is block restricted.

**Comment by dschwen** [comment]
So... talking with @bwspenc for half an hour now, the potential uses for the DiscreteMaterials are getting clearer to me. Unfortunately what came out of this discussion was that a different design might be a better fit here.

One potential road we need to go down is to make the entire stress/strain calculation in TensorMechanics iterative. That would mean making _all_ materials in TensorMechanics DiscreteMaterials and having a Master Material that calls them all. However this would be a breaking change and would complicate the simpler applications of TensorMechanics.

A possible solution would be to give _all_ materials the capability to be DiscreteMaterials! We could have a new toplevel block
```
[DiscreteMaterials]
[]
```
Which materials could get added to resulting in not autoupdating them from MOOSE. Materials in that block could be fetched by `getDiscreteMaterial` (MOOSE can do all the error checking here). This would basically get rid of the specific DiscreteMaterial base class. The recompute methods would be added to the Material base class. Of course these methods would only be available if the getDiscreteMaterial method was called, so everything should be clean and consistent.

This design would allow us to use TensorMechanics as before, _but_ would give us the option to simply move all the stress/strain calculation classed into the `[DiscreteMaterials]` top-level block and add an iterating master Material to the `[Materials]` block.

**Comment by aeslaughter** [comment]
This wouldn't be all the difficult to do, but I would like to see the BISON
team implement what they need first. If the idea of DiscreteMaterials works
for them, then we can work on generalizing the idea.

**Comment by bwspenc** [comment]
@aeslaughter I agree. Let's see if we can do what we need to do with DiscreteMaterial, and then once we have a better handle on how that works for us, we can look at doing something like what Daniel suggested.

**Comment by dschwen** [comment]
Yeah, sounds good. It should be possible to switch to the other design without any major API changes (only input files would have to be adjusted).

**Comment by sapitts** [comment]
@aeslaughter I'm working on a tensor mechanics version of the solid mechanics J2 radial return mapping and will hopefully have something ugly functioning tomorrow as a work in progress demo.

**Comment by lindsayad** [comment]
6 years on I wish we had decided on a different name than `discrete` (I understand that we went down that path before this PR). `discrete` is just an unfortunate name in a repository that's all about discretizing PDEs
